.travis.yml

```diff
@@ -96,7 +96,7 @@ install:
 script:
   - |-
     echo "using $TOOLSET : : $TRAVIS_COMPILER ;" > ~/user-config.jam
-  - ./b2 -j 3 libs/serialization/test toolset=$TOOLSET link=${LINK:-shared}
+  - ./b2 -j 3 libs/serialization/test toolset=$TOOLSET link=${LINK:-shared} cxxflags=-fPIC cflags=-fPIC
```

Comment:
Because I need this when building static boost but link it against dynamic libs to reproduce this. For static boost bjam does not do this

Reply:
FWIW - I never add the flags -fPIC.  I "trust" bam to "do the right thing" for the platform and configuration.  This could well be responsible for getting different test results.  In my view if these are necessary, they should be added to the the appropriate jam script for all shared libraries.  Then there is the issue of what the "link=shared" switch actually does.  It's my belief that it would link to the shared version of the C++ runtime library.  But then there are also C++ system libraries like libc++ - a different thing.  Then there are one's own shared libraries as we are using in our tests here.  So what does link=shared/static apply to?  As far as I'm concerned it's never been explained.  Oh and don't forget the run-link feature.  How should it be set in what circumstances? It goes on an on and on and on.  B2 build system injects myriad unexplained side-effects into one's code.  Worse these side-effects may vary silently from platform to platform.  It also doesn't help at all that the boost test matrix doesn't make it easy to see how these things are set for each test.  All this makes it very, very, very difficult to know that your build is what you really want.  And makes it much, much more difficult to collaborate given that there's no easy way to guarantee that differing testers are actually testing the same thing.

Reply:
> FWIW - I never add the flags -fPIC. I "trust" bam to "do the right thing" for the platform and configuration.

Unfortunately, `bjam` doesn't do the right thing when building shared libraries that incorporate static libraries. It adds `-fPIC` when building the shared library, but not when building the static library.

> Then there is the issue of what the "link=shared" switch actually does. It's my belief that it would link to the shared version of the C++ runtime library.

No, that's `runtime-link=shared`. `link=shared` applies to the libraries built by `b2`, makes them shared.

Reply:
off topic - so what is runtime-link set to by default?  My understanding is that at least on windows if one module is set to use static version of the C++ runtime and another is set to use Dll  version, ceos ensues. as far as I know, there's no obstacle to one doing this.  b2 is opaque to me.

and BTW - it seems that -fPIC shouldn't be necessary when building a static library as the addresses are fixed up at link time.

Reply:
The default for `runtime-link` is `shared`. `b2` will not let you mix the two. `-fPIC` is necessary when building a static library if you later incorporate this library as part of a shared library.

Reply:
@ `-fPIC`: The crash happens when static boost is linked into shared libs. Hence:

- Build a new library: https://github.com/boostorg/serialization/pull/131/files#diff-e97299ad55a4013dc19889043aa464a5R55
- Make it a shared library: https://github.com/boostorg/serialization/pull/131/files#diff-e97299ad55a4013dc19889043aa464a5R60
- Link against static boost: https://github.com/boostorg/serialization/pull/131/files#diff-e97299ad55a4013dc19889043aa464a5R58
- Due to this (static lib is **consumed** by dynamic lib which is not "backpropagated by bjam) add -fPIC

Reply:
I like to add `-fPIC` via a build feature, it's more "surgical" and automatic (you need to pass correct flags on bjam command line).
See here: https://github.com/res2k/serialization/commit/b279e155bb9ba18499c38daa0dfe3a190a23390d